docs/architecture/providers/parameterized.md

```diff
@@ -1,14 +1,46 @@
 (parameterized-providers)=
-# Parameterized providers
+# Parameterized Providers
 
-*Parameterized providers* are a feature of Pulumi that allows a caller to change
-a provider's behaviour at runtime in response to a
-[](pulumirpc.ResourceProvider.Parameterize) call. Where a
-[](pulumirpc.ResourceProvider.Configure) call allows a caller to influence
-provider behaviour at a high level (e.g. by specifying the region in which an
-AWS provider should operate), a [](pulumirpc.ResourceProvider.Parameterize) call
-may change the set of resources and functions that a provider offers (that is,
-its schema). A couple of examples of where this is useful are:
+*Parameterized providers* are a provider feature where a user can generate and
+use a custom SDK based on some custom input.
+
+Key rules:
+
+* Each provider parameterization is run in its own provider instance.
+* Parameterized package names must be unique within the program where its added.
+* Parameterize is always called before Configure.
+* Parameterize is always called before GetSchema.
+
+For a provider to support parameterization it must implement the
+[](pulumirpc.ResourceProvider.Parameterize) call. The
+[](pulumirpc.ParameterizeRequest) will either contain the args passed via the
+CLI from `pulumi package add` or `pulumi package gen-sdk`; or will be passed a
+previous parameterization result (a binary blob). Parameterize will always be
+called before either Configure or GetSchema are called.
+[](pulumirpc.ParameterizeResponse) must return a response with the *package*
+name and version, which must match the generated SDK (from the schema). The name
+of the parameterized package must be unique and will be used as the first part of
+all resource tokens and used to route RegisterResourceRequests back to the
+Parameterized provider instance. The version of the parameterized package can be
+anything the provider would like and does not have to align to the version of the
+base provider. The version does not have any specific purpose right now - it's
+purely informational, but is required to be set.
+
+Once [](pulumirpc.ResourceProvider.Parameterize) has been called (in either mode)
+the engine must be able to call GetSchema and Configure to start to use the
+provider or generate the SDK. For replacement parameterization, once Parameterize
+has been called, the provider will only be used in its parameterized form.
+
+Once parameterized, GetSchema can be called by the engine. The GetSchema request
+will contain the same name and version information as was returned from
+[](pulumirpc.ResourceProvider.Parameterize) in the SubpackageName and
+SubpackageVersion fields, respectively. The generated schema *must* have the same
+name and version as the SubpackageName and SubpackageVersion. GetSchema must be
+callable after either the "args" or "value" mode of parameterize was called. It's
+expected that GetSchema doesn't have to be fast and that the engine will cache
+schemas as required.
```

Comment:
I like this content but I wonder about intermingling the "reference" documentation (what parameterization is, how it works) with "tutorial"-esque documentation (how you should implement it, what you can expect) is muddying some of the water here. I feel like potentially breaking out this kind of content into a section, or putting it in the existing implementer's guide, might be a better fit?

Reply:
I've separated this a bit more within this document. Can always add it somewhere else later and remove from here.